### PR TITLE
Add missing SPV_KHR_ray_tracing storage class VUID

### PR DESCRIPTION
**Pull request by sjfricke**
Currently, there is code in `spirv-val` to check the `Storage Classes` in `SPV_KHR_ray_tracing` and while I am adding support right now, I realized there is duplicate language in the Vulkan spec that is already well documented in the [SPV_KHR_ray_tracing](https://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/master/extensions/KHR/SPV_KHR_ray_tracing.html#_modifications_to_the_spir_v_specification) spec

My argument for the removal is
- The Vulkan Spec was missing `ShaderRecordBufferKHR` so it is not complete in the Vulkan spec currently
- There are much more details missing from the SPIR-V spec missing here (eg. a VU for `HitAttributeKHR` can only be read from `AnyHitKHR` and `ClosestHitKHR`)
- single origin to change in future if another Storage Class interacts with the extension

### CODE REVIEW

**Comment by dgkoch** [comment]
I'd say the downside of not having this in the Vulkan spec is that it'll be harder to find/easier to miss for validation, and there wouldn't be any VUIDs assigned for the checks.  Will discuss in the Ray TSG though.

**Comment by sjfricke** [comment]
At the end of the day, as long as everything is properly being validated in `spirv-val` (which it is currently not so I want to add) I will be happy 😁, so happy with what the WG decides. (I realize this is the not the first time we hit this situation with SPIR-V related VUs 🙃)

The main thing that caught me off guard at first was there was no VU for `ShaderRecordBufferKHR` so I spent time trying to see if there was something subtle that the Vulkan spec disallowed that was a superset of the SPIR-V validation

**Comment by dgkoch** [comment]
> The main thing that caught me off guard at first was there was no VU for `ShaderRecordBufferKHR` so I spent time trying to see if there was something subtle that the Vulkan spec disallowed that was a superset of the SPIR-V validation

Could we interest you in proposing an MR that adds the stuff that is missing from the SPIR-V VU section instead?

**Comment by sjfricke** [comment]
Of course, will line it up with the Validation PR as well 💪

**Comment by sjfricke** [comment]
@dgkoch hopefully this is good now

File: appendices/spirvenv.txt [hunk 1]
```diff
@@ -491,6 +491,8 @@ endif::VK_VERSION_1_1[]
   * [[VUID-{refpage}-HitAttributeKHR-04703]]
     A variable with code:HitAttributeKHR storage class must: only be written
     to in an intersection shader
+  * A variable with code:HitAttributeKHR storage class must: only be read
+    to in an any-hit or closest hit shader
```
Comment (stu-s) [inline appendices/spirvenv.txt:495]:
```suggestion
    from in an any-hit or closest hit shader
```

File: appendices/spirvenv.txt [hunk 2]
```diff
@@ -491,6 +491,8 @@ endif::VK_VERSION_1_1[]
   * [[VUID-{refpage}-HitAttributeKHR-04703]]
     A variable with code:HitAttributeKHR storage class must: only be written
     to in an intersection shader
+  * A variable with code:HitAttributeKHR storage class must: only be read
+    from in an any-hit or closest hit shader
```
Comment (dgkoch) [inline appendices/spirvenv.txt:495]: Also intersection (it can be read after being written). 
Depending on exactly how 4703 is parsed, it could be added up there instead..

Comment (sjfricke) [reply appendices/spirvenv.txt:495]: well "read/written" are taken from the SPIR-V spec

> They can be written to only in IntersectionKHR execution model and read from only in AnyHitKHR and ClosestHitKHR execution models.

but in practice, it turned into `OpLoad`/`OpStore` in the `spirv-val` PR

I guess what is a use case to read the value from the `HitAttributeKHR` variable instead of the variable used to write to it?

Comment (dgkoch) [reply appendices/spirvenv.txt:495]: I don't believe we have anything in SPIR-V that prohibits from reading back output variables, and we definitely want to allow that at the HLL. In theory you could do a compiler transform that does what you suggested, but I doubt that glslang does that, and I think that is getting us off into the weeds here. 

Looking back at the GLSL we say: 

    4.3.X hitAttributeEXT Variables

    These are allowed only in intersection, any-hit, and closest-hit shaders.
    It is a compile-time error to use them in any other stage. They can be read
    in any-hit, and closest-hit shaders. They can be written to and read from
    only in intersection shaders (values are undefined before being written to).

In the SPIR-V we say: 

    HitAttributeKHR
    ...
    Variables declared with this storage class are allowed only in IntersectionKHR,
    AnyHitKHR and ClosestHitKHR execution models. They can be written to only
    in IntersectionKHR execution model and read from only in AnyHitKHR and
    ClosestHitKHR execution models.

So the intent is to clearly allow R/W in Intersection and R in AnyHit and ClosestHit.

Looking back up a few more lines - we have VUID 04701 which already restricts HitAttributeKHR to Intersection, AnyHit and ClosestHit stages.
So 04703 is trying to say that Intersection is the only stage that is allowed to write to HitAttributeKHR (maybe we need slightly different wording to clarify that).
And your new VU should be trying to say that the only stages allowed to read from HitAttributeKHR are Intersection, AnyHit, and Closest hit.

How about: 

  * [[VUID-{refpage}-HitAttributeKHR-04703]]
    A variable with code:HitAttributeKHR storage class must: not be written to in any stage which is not an intersection shader.
  * A variable with code:HitAttributeKHR storage class must: not be read from in any stage which is not an intersection, any-hit, or closest hit shader.

But then the newly proposed one seems pretty redundant with the existing 04701 VU, so I think my only suggestion here might be to try to reword the 04703 VU, so that it's clear that "writing is only permitted in IS" and NOT "is write-only in IS"

Comment (sjfricke) [reply appendices/spirvenv.txt:495]: > They can be written to only in IntersectionKHR execution model and read from only in AnyHitKHR and ClosestHitKHR execution models. 

ok... I slightly misread that and "you can ONLY read from those 2 stages" not it was "read only IN those stages"

|   | Any | Closest | Intersection |
|---|-----|---------|--------------|
| W |     |         | X            |
| R | X   | X       | X            |

I think both the `4703` and `4701` already cover this then already! Lets just remove this line and maybe reword the SPIR-V spec to just say

> They can be written to only in IntersectionKHR execution model.

as adding the "read from only" is not need as it would be implied from the above sentence stating it is allowed in all 3 stages

Comment (sjfricke) [inline appendices/spirvenv.txt:495]:
```suggestion
```

**Review by dgkoch — APPROVED** [review]
LGTM now.

**Review by stu-s — APPROVED** [review]
LGTM

**Comment by dgkoch** [comment]
@oddhack please pick this up for the next spec update.
